### PR TITLE
feat: make the query params extensible

### PR DESCRIPTION
**Pull request by iamskok**
As discussed in https://github.com/gatsbyjs/gatsby/issues/5030#issuecomment-453753641 I have made a minor update to params parsing logic, which is supposed to avoid conflicts between related `gatsby-remark-*` plugins, e.g. [gatsby-remark-code-buttons](https://github.com/iamskok/gatsby-remark-code-buttons)

### CODE REVIEW

**Comment by DSchau** [comment]
Hello! I’m out on vacation until next Thursday—but I’ll take a look at this promptly when I’m back!

**Comment by iamskok** [comment]
@DSchau looking forward to it. Have a great time 👍

**Comment by iamskok** [comment]
Hi, Dustan! 

What do you think about my PR?

File: src/index.js [hunk 1]
```diff
@@ -1,21 +1,32 @@
...
+  { markdownAST },
+  { className: customClassName }
+) {
   visit(markdownAST, 'code', (node, index) => {
-    const [language, query] = (node.lang || '').split(':');
-    const { title } = qs.parse(query);
+    const [language, ...params] = (node.lang || '').split(':');
+    const query = params.join('&');
+    const options = qs.parse(query);
+    const title = options.title;
```
Comment (DSchau) [inline src/index.js]: What about:

```suggestion
    const { title, ...rest } = options;
```

and then you can use ...rest later to re-construct?

File: src/index.js [hunk 2]
```diff
@@ -1,21 +1,32 @@
...
     if (!title || !language) {
       return;
     }
 
+    delete options['title'];
+    let newQuery = '';
```
Comment (DSchau) [inline src/index.js:16]: how about

```js
const newQuery = `:` + Object.keys(rest).map(key => `${key}=${rest[key]}`).join('&')
```

**Review by DSchau — COMMENTED** [review]
Left some comments; thanks!

File: src/index.js [hunk 3]
```diff
@@ -9,15 +9,18 @@ module.exports = function gatsbyRemarkCodeTitles(
...
+    const { title, ...rest } = options;
     if (!title || !language) {
       return;
     }
 
-    delete options['title'];
     let newQuery = '';
-    for (let key in options) {
-      newQuery += `:${key}=${options[key]}`;
+    if (Object.keys(rest).length) {
```
Comment (DSchau) [inline src/index.js]: rest will always be an object, and the empty case will be handled already. Not needed.

**Comment by iamskok** [comment]
Everything is updated

**Review by DSchau — COMMENTED** [review]
@iamskok have you been able to validate this with a real project?

In looking at it, I'm not sure this will work. We directly augment the original AST, so subsequent plugins will still see the lang, but they'll be parsing a different structure (what is now HTML) rather than a code snippet, which won't work regardless.

Also - in thinking about it, this seems like a more natural syntax:

````
```js:title=hello-world.js&clipboard=false
alert('hello world')
```
````

which actually doesn't work (the parser fails on the & and parses it into unicode).

I added a few tests to your branch (some skipped intentionally) so you can see the current behavior.

Thoughts?

**Comment by iamskok** [comment]
> @iamskok have you been able to validate this with a real project?

Yes, it seemed to work fine on one of my side projects.

> We directly augment the original AST, so subsequent plugins will still see the lang, but they'll be parsing a different structure (what is now HTML) rather than a code snippet, which won't work regardless.

Can you provide a little more explanation? Not sure what you meant by saying that subsequent plugins will still see the lang, but they'll be parsing a different structure (what is now HTML). 
The object with `html` type will be added before the object with `code` type. Will it somehow prevent other plugins from working properly?  It looks like we are still filtering AST by `code` type. 

```js
// Before
{ type: 'root',
      children:
       [ { type: 'code',
           lang: 'js:title=hello-world.js&clipboard=true',
           meta: null,
           value: "alert('oh shit waddup')",
           position: [Position] } ],
      position:
       { start: { line: 1, column: 1, offset: 0 },
         end: { line: 5, column: 5, offset: 75 } } }

// After
{ type: 'root',
      children:
       [ { type: 'html',
           value: '<div class="gatsby-code-title">hello-world.js</div>' },
         { type: 'code',
           lang: 'js:clipboard=true',
           meta: null,
           value: "alert('oh shit waddup')",
           position: [Object] } ],
      position:
       { start: { line: 1, column: 1, offset: 0 },
         end: { line: 5, column: 5, offset: 75 } } }
```

> which actually doesn't work (the parser fails on the & and parses it into unicode).

Yes, I got the same problem with `&copy` param being parsed as `©`, so I decided to change it to `clipboard`.  The skipped test is going to pass if `copy` will be switched to `clipboard`.

**Comment by DSchau** [comment]
@iamskok want to make that test pass and then we can get this merged/published?

**Comment by iamskok** [comment]
Hey, @DSchau 👋 

Did you have a chance to look at the test?

**Pull request merged**